### PR TITLE
Resolve the last failing doctests, add doctests to CI

### LINKED ISSUE #2989: Failing tests in DocStrings
Checking the Python files in NLTK with "python -m doctest" reveals that many tests are failing. In many cases, the failures are just cosmetic discrepancies between the expected and the actual output, such as missing a blank line, or unescaped linebreaks. Other cases may be real bugs.

If these failures could be avoided, it would become possible to improve CI by running "python -m doctest" each time a python file was modified.

Here is an updated list, after applying #2990, and unzipping those corpora marked unzip="1":

> nltk/parse/corenlp.py: 9 passed and 33 failed.
nltk/parse/stanford.py: 1 passed and 16 failed.
nltk/corpus/reader/framenet.py: 79 passed and 9 failed.
nltk/tokenize/regexp.py: 14 passed and 7 failed.
nltk/draw/table.py: 0 passed and 6 failed.
nltk/tokenize/simple.py: 7 passed and 6 failed.
nltk/text.py: 16 passed and 5 failed.
nltk/tokenize/stanford_segmenter.py: 3 passed and 5 failed.
nltk/draw/util.py: 1 passed and 4 failed.
nltk/tag/hunpos.py: 1 passed and 4 failed.
nltk/tokenize/__init__.py: 6 passed and 4 failed.
nltk/translate/ibm1.py: 12 passed and 4 failed.
nltk/tag/brill_trainer.py: 23 passed and 3 failed.
nltk/tokenize/destructive.py: 8 passed and 3 failed.
nltk/tokenize/treebank.py: 32 passed and 3 failed.
nltk/tokenize/util.py: 27 passed and 3 failed.
nltk/translate/meteor_score.py: 8 passed and 3 failed.
nltk/corpus/reader/categorized_sents.py: 5 passed and 2 failed.
nltk/corpus/reader/opinion_lexicon.py: 3 passed and 2 failed.
nltk/corpus/reader/reviews.py: 8 passed and 2 failed.
nltk/metrics/agreement.py: 8 passed and 2 failed.
nltk/parse/dependencygraph.py: 13 passed and 2 failed.
nltk/parse/evaluate.py: 6 passed and 2 failed.
nltk/parse/util.py: 5 passed and 2 failed.
nltk/tag/stanford.py: 2 passed and 2 failed.
nltk/tokenize/stanford.py: 3 passed and 2 failed.
nltk/corpus/__init__.py: 1 passed and 1 failed.
nltk/corpus/reader/comparative_sents.py: 5 passed and 1 failed.
nltk/corpus/reader/__init__.py: 1 passed and 1 failed.
nltk/corpus/reader/pros_cons.py: 2 passed and 1 failed.
nltk/probability.py: 39 passed and 1 failed.
nltk/sentiment/util.py: 7 passed and 1 failed.
nltk/stem/arlstem2.py: 3 passed and 1 failed.
nltk/stem/rslp.py: 3 passed and 1 failed.
nltk/stem/snowball.py: 6 passed and 1 failed.
nltk/tag/__init__.py: 12 passed and 1 failed.
nltk/tokenize/casual.py: 9 passed and 1 failed.
nltk/translate/ibm2.py: 18 passed and 1 failed.
nltk/translate/ibm3.py: 23 passed and 1 failed.
nltk/translate/ibm4.py: 25 passed and 1 failed.
nltk/translate/ibm5.py: 22 passed and 1 failed.

### PR DESCRIPTION
**Pull request by tomaarsen**
Resolves #2989

Hello!

## Pull request overview
* Fix all tests from this list:
    ```
    ====================================================================================== short test summary info ====================================================================================== 
    FAILED nltk/text.py::nltk.text.Text.collocation_list
    FAILED nltk/draw/table.py::nltk.draw.table.MultiListbox.__init__
    FAILED nltk/draw/table.py::nltk.draw.table.MultiListbox.configure
    FAILED nltk/draw/table.py::nltk.draw.table.Table
    FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.CoreNLPDependencyParser
    FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.CoreNLPParser
    FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.GenericCoreNLPParser.tag
    FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.GenericCoreNLPParser.tokenize
    FAILED nltk/parse/transitionparser.py::nltk.parse.transitionparser.demo
    FAILED nltk/tokenize/punkt.py::nltk.tokenize.punkt.PunktSentenceTokenizer._match_potential_end_contexts
    ========================================================== 10 failed, 689 passed, 29 skipped, 9 xfailed, 22 warnings in 143.19s (0:02:23) ===========================================================
    ```
* Add automatic doctests to the CI, making us go from 489 to 686 automatically passing tests 🎉
* Update `requirements-ci.txt` with `click`, `python-crfsuite` and `tqdm` as they are required for doctests.
* Skip additional doctests for `tkinter`, as `tkinter.Tk()` requires a display to exist.
* Replace `OSError` with `LookupError` when SENNA can't be found.
* Resolve breaking issue in `MultiListBox` that prevents it from being instantiated.
* Update the Stanford CoreNLP version used in the CI from `full 2018-10-05` to the most recent `4.5.1`.
* Void the third party CI cache every time `tools/github_actions/third-party.sh` is updated.
* Add a `check_jar` function which calls `pytest.skip` if a jar is unavailable, and otherwise does nothing.

### Details
The changes are primarily very small and few and far between, so I'm going to add comments to the changed code rather than describing the changes in this PR text itself. That way, it's easier to see which comment refers to what piece of code. The only exception is to CoreNLP, which is a somewhat big change:

#### CoreNLP
I've reworked the doctests in that file to include `server = CoreNLPServer(...)`, `server.start()`, `server.stop()` or `with CoreNLPServer() as server:`. That way, the tests actually do pass now if the CoreNLP jars are on `CLASSPATH`.

Thanks to @ekaf and @Madnex for working on the other doctests & compiling the broken ones.

- Tom Aarsen

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -125,4 +125,4 @@ jobs:
       - name: Run pytest
         shell: bash
         run: |
-          pytest --numprocesses auto -rsx --doctest-modules nltk/test
+          pytest --numprocesses auto -rsx --doctest-modules nltk
```
Comment (tomaarsen) [inline .github/workflows/ci.yaml:128]: `--doctest-modules` was already always on, but it didn't do much, as it only looked in `nltk/test`. Looking in `nltk` directly was broken until #3033, but it works now!

File: nltk/classify/senna.py [hunk 2]
```diff
@@ -30,9 +30,9 @@
 Note: Unit tests for this module can be found in test/unit/test_senna.py
 
 >>> from nltk.classify import Senna
->>> pipeline = Senna('/usr/share/senna-v3.0', ['pos', 'chk', 'ner'])
+>>> pipeline = Senna('/usr/share/senna-v3.0', ['pos', 'chk', 'ner'])  # doctest: +SKIP
```
Comment (tomaarsen) [inline nltk/classify/senna.py:33]: All `Senna(...)` sections are skipped, as they break on the Windows and Mac CI's, because the `SENNA` environment variable [does exist](https://github.com/nltk/nltk/blob/develop/.github/workflows/ci.yaml#L11), but points to a path that doesn't.

File: nltk/classify/senna.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def __init__(self, senna_path, operations, encoding="utf-8"):
                 self._path = path.normpath(environ["SENNA"]) + sep
                 exe_file_2 = self.executable(self._path)
                 if not path.isfile(exe_file_2):
-                    raise OSError(
+                    raise LookupError(
```
Comment (tomaarsen) [inline nltk/classify/senna.py:65]: I believe that `LookupError` is more logical here. We also use `LookupError` in other cases where looking up a certain resource fails.

File: nltk/corpus/reader/markdown.py [hunk 4]
```diff
@@ -206,6 +204,8 @@ def concatenated_view(self, reader, fileids, categories):
         )
 
     def metadata_reader(self, stream):
+        from yaml import safe_load
+
```
Comment (tomaarsen) [inline nltk/corpus/reader/markdown.py:208]: This way we don't need to add `pyyaml` to the requirements unnecessarily.

File: nltk/draw/table.py [hunk 5]
```diff
@@ -76,13 +76,8 @@ def __init__(self, master, columns, column_weights=None, cnf={}, **kw):
...
             to configure all listboxes.  E.g.:
-                >>> from nltk.draw.util import TextWidget, CanvasFrame, SpaceWidget
-                >>> from nltk.draw.table import MultiListbox
-                >>> cf = CanvasFrame(closeenough=10, width=300, height=300)
-                >>> c = cf.canvas()
-                >>> master = TextWidget(c, "hiya there", draggable=1)
-                >>> master = SpaceWidget(c, 0, 30)
-                >>> mlb = MultiListbox(master, 5, label_foreground='red')
+                >>> root = Tk()  # doctest: +SKIP
+                >>> MultiListbox(root, ["Subject", "Sender", "Date"], label_foreground='red').pack()  # doctest: +SKIP
```
Comment (tomaarsen) [inline nltk/draw/table.py:80]: The old code didn't really make sense: I thought `master` or `root` was always a `tkinter.Tk()` instance. Beyond that, it didn't compile I believe.

I also have to skip these, as `Tk()` fails on a device without a display.

File: nltk/draw/table.py [hunk 6]
```diff
@@ -139,7 +134,7 @@ def __init__(self, master, columns, column_weights=None, cnf={}, **kw):
             # the default listbox behavior, which scrolls):
             lb.bind("<B1-Leave>", lambda e: "break")
             # Columns can be resized by dragging them:
-            l.bind("<Button-1>", self._resize_column)
+            lb.bind("<Button-1>", self._resize_column)
```
Comment (tomaarsen) [inline nltk/draw/table.py:137]: This was a breaking bug that prevented `MultiListBox` from being instantiated. It has been in here for years. I suppose nobody uses it?

File: nltk/parse/transitionparser.py [hunk 7]
```diff
@@ -752,6 +752,7 @@ def demo():
     >>> parser_std.train([gold_sent],'temp.arcstd.model', verbose=False)
      Number of training examples : 1
      Number of valid (projective) examples : 1
+    >>> input_file.close()
```
Comment (tomaarsen) [inline nltk/parse/transitionparser.py:755]: Temporary files were being removed before they were closed, causing exceptions (and leftover files on my device...)

File: nltk/text.py [hunk 8]
```diff
@@ -270,8 +270,7 @@ def findall(self, regexp):
         a single token must be surrounded by angle brackets.  E.g.
 
         >>> from nltk.text import TokenSearcher
-        >>> print('hack'); from nltk.book import text1, text5, text9
-        hack
+        >>> from nltk.book import text1, text5, text9
```
Comment (tomaarsen) [inline nltk/text.py:273]: The hack is unnecessary here. The tests pass like this, too.

File: nltk/text.py [hunk 9]
```diff
@@ -410,20 +409,7 @@ def collocation_list(self, num=20, window_size=2):
...
-            text1: Moby Dick by Herman Melville 1851
-            text2: Sense and Sensibility by Jane Austen 1811
-            text3: The Book of Genesis
-            text4: Inaugural Address Corpus
-            text5: Chat Corpus
-            text6: Monty Python and the Holy Grail
-            text7: Wall Street Journal
-            text8: Personals Corpus
-            text9: The Man Who Was Thursday by G . K . Chesterton 1908
+            >>> from nltk.book import text4
```
Comment (tomaarsen) [inline nltk/text.py:412]: This one is a bit tricky. `nltk.book` prints some stuff when imported, which is what we were testing for. However, it only prints it the first time it is imported.
if you do `pytest --doctest-modules nltk/text.py`, then it imports `nltk.book` for the first time here, and then it expects some printed output, but if you do `pytest --doctest-modules ./nltk`, then it imports it elsewhere differently (probably in some pytest code that traverses everything under `./nltk`), and then we expect this import to print nothing.

Long story short: the new situation passes on `pytest --doctest-modules nltk` (which is the important case), but fails on `pytest --doctest-modules nltk/text.py` (which is the unimportant case).

File: nltk/tokenize/punkt.py [hunk 10]
```diff
@@ -1396,12 +1396,13 @@ def _match_potential_end_contexts(self, text: str) -> Iterator[Tuple[Match, str]
...
-                last_space_index += previous_slice.stop
+            index_after_last_space = self._get_last_whitespace_index(before_text)
+            if index_after_last_space:
+                # + 1 to exclude the space itself
+                index_after_last_space += previous_slice.stop + 1
             else:
-                last_space_index = previous_slice.start
-            prev_word_slice = slice(last_space_index, match.start())
+                index_after_last_space = previous_slice.start
+            prev_word_slice = slice(index_after_last_space, match.start())
```
Comment (tomaarsen) [inline nltk/tokenize/punkt.py:1405]: This method, which is supposed to return a "context" word that comes before a specific token of interest, used to also return the whitespace that preceded that word. I've fixed this now, so it doesn't return the whitespace anymore. It should not affect the tokenization, as that already `strip()`-ed away the extra whitespace anyways.

File: requirements-ci.txt [hunk 11]
```diff
@@ -1,8 +1,11 @@
...
 gensim>=4.0.0
 matplotlib
 pytest
 pytest-mock
 pytest-xdist[psutil]
+python-crfsuite
 regex
 scikit-learn
+tqdm
 twython
```
Comment (tomaarsen) [inline requirements-ci.txt:11]: These are (now) required to run the CI due to the doctests.

**Comment by tomaarsen** [comment]
I'm going to make some small changes to how the CoreNLP tests are skipped. Currently, they are all skipped in the CI, which I also want to change.

**Comment by ekaf** [comment]
This looks good!

```
pytest -vv --doctest-modules --doctest-continue-on-failure ./nltk/
```


```
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-7.1.2, pluggy-1.0.0 -- /usr/bin/python3
[...]
===== 687 passed, 41 skipped, 9 xfailed, 22 warnings in 417.37s (0:06:57) ======

```

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (ekaf) [reply .github/workflows/ci.yaml:128]: Thanks @tomaarsen, this should make CI much more robust. I did not expect that it would be ready so soon!

Comment (stevenbird) [reply .github/workflows/ci.yaml:128]: Oh!

File: nltk/draw/table.py (hunk 6 above)
Comment (stevenbird) [reply nltk/draw/table.py:137]: Wow... yes, this is telling, isn't it

File: nltk/text.py (hunk 9 above)
Comment (stevenbird) [reply nltk/text.py:412]: Agreed

**Comment by stevenbird** [comment]
Nice work @tomaarsen! Thanks @ekaf.
